clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java

```diff
@@ -0,0 +1,139 @@
+             * We want to return the future to the map, but we have to return a map at this point.
+             * We need to dereference the future while propagating any exception.
+             */
+            return valueFromFuturePropagatingExceptionsAsUnchecked(mapFuture);
+        });
+    }
+
+    /**
+     *
+     * @return a future indicating the distinct users that were requested (either explicitly or implicitly via
+     * describe-all).  The future will not complete successfully if the user is not authorized to perform the describe
+     * operation; otherwise, it will complete successfully as long as the list of users with credentials can be
+     * successfully determined within some hard-coded timeout period.
+     */
+    public KafkaFuture<List<String>> users() {
+        return usersFuture;
+    }
+
+    /**
+     *
+     * @param userName the name of the user description being requested
+     * @return a future indicating the description results for the given user. The future will complete exceptionally if
+     * the future returned by {@link #users()} completes exceptionally.  If the given user does not exist in the list
+     * of requested users then the future will complete exceptionally with
+     * {@link org.apache.kafka.common.errors.ResourceNotFoundException}.
+     */
+    public KafkaFuture<UserScramCredentialsDescription> description(String userName) {
+        KafkaFuture<Void> succeedsOnlyIfUsersFutureSucceeds = KafkaFuture.allOf(usersFuture);
+        return succeedsOnlyIfUsersFutureSucceeds.thenApply(void1 -> {
+            // it is possible that there is no future for this user (for example, the original describe request was for
+            // users 1, 2, and 3 but this is looking for user 4), so explicitly take care of that case
+            KafkaFuture<UserScramCredentialsDescription> requestedUserFuture = perUserFutures.get(userName);
+            if (requestedUserFuture == null) {
+                throw new ResourceNotFoundException("No such user: " + userName);
+            }
+            KafkaFuture<Void> succeedsOnlyIfRequestedUserFutureSucceeds = KafkaFuture.allOf(requestedUserFuture);
+            KafkaFuture<UserScramCredentialsDescription> descriptionFuture = succeedsOnlyIfRequestedUserFutureSucceeds.thenApply(void2 ->
+                valueFromFutureGuaranteedToSucceedAtThisPoint(requestedUserFuture));
+            /* At this point it is only the users future that is guaranteed to have succeeded.
+             * We want to return the future to the description, but we have to return a description at this point.
+             * We need to dereference the future while propagating any exception.
+             */
+            return valueFromFuturePropagatingExceptionsAsUnchecked(descriptionFuture);
+        });
+    }
+
+    private static <T> T valueFromFutureGuaranteedToSucceedAtThisPoint(KafkaFuture<T> future) {
+        try {
+            return future.get();
```

Comment:
You should not be calling `get()` here.

In general it seems like what you really want is a single future behind the scenes that returns everything that you fetched from your single RPC.  Then you just need a few translation functions that pull out the part you need for each future that you're returning to the user.

Reply:
>  a single future behind the scenes that returns everything that you fetched from your single RPC 

That would introduce an assumption that none of the users' results are available until all of the users' results are available.  For example, currently this class accepts two arguments in its constructor: a future to a list of user names, and a map of usernames-to-futures for the results of each user.  If instead for the second argument we accept a single future to a map of usernames-to-results, then that future must succeed all the time (because we need to distinguish between successes and failures for different users).  Each user's result would have to contain the error code, and no user's results will be available to the client until all of the users' results (the underlying map) is available.  It is true that this is how it is implemented today -- no user's results will be available until all users' results are available, and each of the individual results has an error code -- but Is it acceptable to impose such an all-or-nothing result availability constraint?  The idea of flexibility in the API going forward is an ambiguous one, so we should explicitly decide here whether to keep it the way it is now (allowing different users to complete at different times even though it isn't possible right now) vs. changing the constructor and constraining the results for any particular user to be available only after the results for all users are available.  What do you think?

Reply:
> You should not be calling get() here

Looking at the implementation of `KafkaFutureImpl` I agree we should not be calling `get()` because I think it means when the users future completes the thread that completes it will wait until the results future for that user also completes.  If this is indeed the case then we definitely do not want to be calling `get()` in our code.  One way to make a hard dependency on the users future completing without causing the admin client thread to block as just described is to provide a 2-level results hierarchy as was previously implemented.  This forces the user to complete 2 futures and to do it in their own thread.

I wonder if we have the constructor accept 2 futures rather than 1+N futures we would still create a dependency between the 2 that would block the admin client thread.  Perhaps the second future would have to already have a dependency on the users future built into it.

One thing we have to be careful about is the constraints we put on the arguments to the constructor -- those constraints, whatever they end up being, have to be clearly documented.  If we don't have to impose any -- then that's great.  But if we have to impose/document some, then we should compare that with the 2-level results hierarchy I had provided earlier -- that imposes no constraints whatsoever.